src/machines/java/com/enderio/machines/common/recipe/SoulBindingRecipe.java

```diff
@@ -0,0 +1,209 @@
+    private final int energy;
+
+
+    public SoulBindingRecipe(ResourceLocation id, Item output, List<Ingredient> inputs, int energy, int exp, @Nullable ResourceLocation entityType) {
+        this.id = id;
+        this.output = output;
+        this.inputs = inputs;
+        this.energy = energy;
+        this.exp = exp;
+        this.entityType = entityType;
+    }
+
+    @Override
+    public int getEnergyCost(SoulBindingRecipe.Container container) {
+        return energy;
+    }
+
+    public int getExpCost() {
+        return exp;
+    }
+
+    @Override
+    public List<OutputStack> craft(SoulBindingRecipe.Container container) {
+        ItemStack vial = container.getItem(0);
+        List<OutputStack> results = getResultStacks();
+        results.forEach(o -> {
+            ItemStack result = o.getItem(); //TODO will this auto update since the stack is updated?
+            vial.getCapability(EIOCapabilities.ENTITY_STORAGE).ifPresent(inputEntity -> {
+                result.getCapability(EIOCapabilities.ENTITY_STORAGE).ifPresent(resultEntity -> {
+                    resultEntity.setStoredEntityData(inputEntity.getStoredEntityData());
+                });
+            });
+        });
+        return results;
+    }
+
+    @Override
+    public List<OutputStack> getResultStacks() {
+        return List.of(OutputStack.of(new ItemStack(output, 1)));
+    }
+
+    @Override
+    public boolean matches(SoulBindingRecipe.Container container, Level pLevel) {
+        container.setNeededXP(0);
+        for (int i = 0; i < inputs.size(); i++) { //Items match
+            if (!inputs.get(i).test(container.getItem(i)))
+                return false;
+        }
+        LazyOptional<IEntityStorage> capability = container.getItem(0).getCapability(EIOCapabilities.ENTITY_STORAGE);
+        if (!capability.isPresent()) { //vial (or other entity storage
+            return false;
+        }
+        if (entityType == null) { //type doesn't matter
+            container.setNeededXP(exp);
+            return ExperienceUtil.getLevelFromFluid(container.getFluidTank().getFluidAmount()) >= exp;
+        }
+        IEntityStorage storage = capability.resolve().get();
+        if (storage.hasStoredEntity() && storage.getStoredEntityData().getEntityType().get().equals(entityType)) { //type matters
+            container.setNeededXP(exp);
+            return ExperienceUtil.getLevelFromFluid(container.getFluidTank().getFluidAmount()) >= exp;
+        }
+        return false;
+    }
+
+    @Override
+    public ResourceLocation getId() {
+        return this.id;
+    }
+
+    @Override
+    public RecipeSerializer<?> getSerializer() {
+        return MachineRecipes.SOUL_BINDING.serializer().get();
+    }
+
+    @Override
+    public RecipeType<?> getType() {
+        return MachineRecipes.SOUL_BINDING.type().get();
+    }
+
+    public static class Container extends RecipeWrapper {
+
+        private final FluidTank fluidTank;
```

Comment:
You may want to swap this for a machineFluidTank, since you only want fluid xp to flow in. You can do both of those things automatically through that class ;)

Reply:
This container is purely used for the recipe and is just a reference to the real tank. But you do raise a good point of having a second look at the real tank, as it might be replaced by machineFluidTank. It also seems to miss somethings right now, good catch.